osidb/tests/test_flaw.py

```diff
@@ -1194,3 +1207,114 @@ def test_flaw_affects_rhscl_invalid_collection(
+        self, affectedness, resolution, should_raise
+    ):
+        """
+        Test that error is raised if any flaw's affect have a
+        invalid combination of affectedness and resolution
+        """
+        flaw = FlawFactory()
+        affect = AffectFactory.build(
+            flaw=flaw, affectedness=affectedness, resolution=resolution
+        )
+        affect.save(raise_validation_error=False)
+        if should_raise:
+            with pytest.raises(
+                ValidationError,
+                match=f"{affect.resolution} is not a valid resolution for {affect.affectedness}.",
+            ):
+                flaw.save()
+        else:
+            assert flaw.save() is None
+
+    @pytest.mark.parametrize(
+        "affectedness,resolution,should_raise",
+        [
+            (Affect.AffectAffectedness.NEW, None, False),
+            (Affect.AffectAffectedness.NEW, Affect.AffectResolution.DEFER, False),
+            (Affect.AffectAffectedness.NEW, Affect.AffectResolution.WONTFIX, False),
+            (Affect.AffectAffectedness.NEW, Affect.AffectResolution.OOSS, False),
+            (Affect.AffectAffectedness.AFFECTED, Affect.AffectResolution.FIX, False),
+            (Affect.AffectAffectedness.AFFECTED, Affect.AffectResolution.DEFER, False),
+            (
+                Affect.AffectAffectedness.AFFECTED,
+                Affect.AffectResolution.DELEGATED,
+                False,
+            ),
+            (
+                Affect.AffectAffectedness.AFFECTED,
+                Affect.AffectResolution.WONTREPORT,
+                False,
+            ),
+            (
+                Affect.AffectAffectedness.AFFECTED,
+                Affect.AffectResolution.WONTFIX,
+                False,
+            ),
+            (Affect.AffectAffectedness.AFFECTED, Affect.AffectResolution.OOSS, False),
+            (Affect.AffectAffectedness.NOTAFFECTED, None, False),
+            (Affect.AffectAffectedness.NEW, Affect.AffectResolution.FIX, True),
+            (Affect.AffectAffectedness.AFFECTED, Affect.AffectResolution.NOVALUE, True),
+            (
+                Affect.AffectAffectedness.NOTAFFECTED,
+                Affect.AffectResolution.DEFER,
+                True,
+            ),
+        ],
+    )
+    def test_validate_affect_status_resolution(
+        self, affectedness, resolution, should_raise
+    ):
+        """
+        Test that error is raised if affect have a invalid
+        combination of affectedness and resolution
+        """
+        flaw = FlawFactory()
+        affect = AffectFactory.build(
+            flaw=flaw, affectedness=affectedness, resolution=resolution
+        )
+        if should_raise:
+            with pytest.raises(
+                ValidationError,
+                match=f"{affect.resolution} is not a valid resolution for {affect.affectedness}.",
+            ):
+                affect.save()
+        else:
+            assert affect.save() is None
```

Comment:
This is almost 1:1 identical to the code above, so this is one of the few cases were a cartesian product of `parametrize` makes sense [1]

[1] https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions (see end of the section for an example)

Reply:
What you said makes sense, but I have submitted a different fixture for this that cover tests for this same validation from both Flaw and Affect point of view in the same function without using the stacking parameters for cartesian product, please take a look and if you still think it is better use the cartesian I will submit a new fix. (I have decided to do this way because otherwise my test would look pretty much the same as my validation, accessing the same variables / doing the same conditions like the below example):
"""
@pytest.mark.parametrize(
    "resolution",
    [
        Affect.AffectResolution.NOVALUE,
        Affect.AffectResolution.FIX,
        Affect.AffectResolution.DEFER,
        Affect.AffectResolution.WONTFIX,
        Affect.AffectResolution.OOSS,
        Affect.AffectResolution.DELEGATED,
        Affect.AffectResolution.WONTREPORT,
    ]
)
@pytest.mark.parametrize(
    "affectedness",
    [
        Affect.AffectAffectedness.NOVALUE,
        Affect.AffectAffectedness.NEW,
        Affect.AffectAffectedness.AFFECTED,
        Affect.AffectAffectedness.NOTAFFECTED,
    ]
)
def test_validate_flaw_affects_status_resolution(
    self, affectedness, resolution
):
    flaw = FlawFactory()
    affect = AffectFactory.build(
        flaw=flaw, affectedness=affectedness, resolution=resolution
    )
    affect.save(raise_validation_error=False)
    if resolution not in AFFECTEDNESS_VALID_RESOLUTIONS[self.affectedness]:
        with pytest.raises(
            ValidationError,
            match=f"{affect.resolution} is not a valid resolution for {affect.affectedness}.",
        ):
            flaw.save()
        with pytest.raises(
            ValidationError,
            match=f"{affect.resolution} is not a valid resolution for {affect.affectedness}.",
        ):
            affect.save()
    else:
        assert affect.save() is None
        assert flaw.save() is None
"""

Reply:
My bad, I should've been more precise, I meant something like:

```python
@parametrize("entity", ["flaw", "affect"])
@parametrize(...)
def test(...):
    ...
    entity = affect if entity == "affect" else flaw
    ...
    assert entity.save() is None
    # or
    with pytest.raises(...):
        entity.save()
```

Reply:
Sorry, now I got your point, I will submit a fixture.